flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/async/queue/StreamRecordQueueEntry.java

```diff
@@ -21,65 +21,55 @@
 import org.apache.flink.annotation.Internal;
 import org.apache.flink.streaming.api.functions.async.AsyncFunction;
 import org.apache.flink.streaming.api.functions.async.ResultFuture;
+import org.apache.flink.streaming.api.operators.TimestampedCollector;
 import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.flink.util.Preconditions;
+
+import javax.annotation.Nonnull;
 
 import java.util.Collection;
-import java.util.concurrent.CompletableFuture;
 
 /**
  * {@link StreamElementQueueEntry} implementation for {@link StreamRecord}. This class also acts
  * as the {@link ResultFuture} implementation which is given to the {@link AsyncFunction}. The
  * async function completes this class with a collection of results.
  *
- * @param <OUT> Type of the asynchronous collection result
+ * @param <OUT> Type of the asynchronous collection result.
  */
 @Internal
-public class StreamRecordQueueEntry<OUT> extends StreamElementQueueEntry<Collection<OUT>>
-	implements AsyncCollectionResult<OUT>, ResultFuture<OUT> {
-
-	/** Timestamp information. */
-	private final boolean hasTimestamp;
-	private final long timestamp;
-
-	/** Future containing the collection result. */
-	private final CompletableFuture<Collection<OUT>> resultFuture;
+class StreamRecordQueueEntry<OUT> implements StreamElementQueueEntry<OUT> {
+	private Collection<OUT> completed;
 
-	public StreamRecordQueueEntry(StreamRecord<?> streamRecord) {
-		super(streamRecord);
+	@Nonnull
+	private final StreamRecord<?> inputRecord;
 
-		hasTimestamp = streamRecord.hasTimestamp();
-		timestamp = streamRecord.getTimestamp();
-
-		resultFuture = new CompletableFuture<>();
+	StreamRecordQueueEntry(@Nonnull StreamRecord<?> inputRecord) {
+		this.inputRecord = Preconditions.checkNotNull(inputRecord);
 	}
 
 	@Override
-	public boolean hasTimestamp() {
-		return hasTimestamp;
+	public boolean isDone() {
+		return completed != null;
 	}
 
+	@Nonnull
 	@Override
-	public long getTimestamp() {
-		return timestamp;
+	public StreamRecord<?> getInputElement() {
+		return inputRecord;
 	}
 
 	@Override
-	public Collection<OUT> get() throws Exception {
-		return resultFuture.get();
-	}
+	public void emitResult(TimestampedCollector<OUT> output) {
+		Preconditions.checkState(completed != null, "Not done yet");
 
-	@Override
-	protected CompletableFuture<Collection<OUT>> getFuture() {
-		return resultFuture;
+		output.setTimestamp(this.inputRecord);
+		for (OUT r : completed) {
+			output.collect(r);
+		}
 	}
 
 	@Override
 	public void complete(Collection<OUT> result) {
-		resultFuture.complete(result);
-	}
-
-	@Override
-	public void completeExceptionally(Throwable error) {
-		resultFuture.completeExceptionally(error);
+		this.completed = result;
```

Comment:
ditto: `this`. Do we need to `checkNotNull(result)`?

Reply:
Added a check to AsyncWaitOperator where we receive use input/